### DISCUSSION TITLE
v1.2.0 - Large update with fixes and new features

### DISCUSSION THREAD

**Discussion by mpaperno** [mpaperno/WASimCommander#18] [Announcements]
### WASimModule
* Fix binary data representation in results for named variable requests with 1-4 byte integer value sizes (`int8` - `int32` types) -- the result data would be encoded as a float type instead. ([8c7724e6])
* Restore ability to use Unit type specifiers when setting and getting Local vars. ([e16049ac])
* Added ability to specify/set a default 'L' var value and unit type in `GetCreate` command to use if the variable needs to be created. ([61a52674])
* `GetCreate` and `SetCreate` commands for non-L types now silently fall back to `Get` and `Set` respectively. ([61a52674])
* Fixed that command response for `GetCreate` was always sent as if responding to a `Get` command. ([61a52674])
* Added `requestId` to error logging and response output for data requests and add more info for `Get` command errors. ([17791eef])
* Added ability to return string type results for Get commands and Named data requests by converting them to calculator expressions automatically on the server. ([983e7ab6])
* Improved automatic conversion to calc code for other variable types by including the unit type, if given, and narrowing numeric results to integer types if needed. ([983e7ab6])
* Prevent possible simulator hang on exit when quitting with active client(s) connections. ([70e0ef31])
* Event loop processing is now paused/restarted also based on whether any connected client(s) have active data requests and if they are paused or not (previously it was only based on if any clients were connected at all). ([90242ed4])
* Fixes a logged SimConnect error when trying to unsubscribe from the "Frame" event (cause unknown). ([90242ed4])
* Data requests with `Once` type update period are now queued if data updates are paused when the request is submitted. These requests will be sent when/if updates are resumed again by the client. Fixes that data would be sent anyway when the request is initially submitted, even if updates are paused. ([fe99bbb2])
* Update reference list of KEY events and aliases as of MSFS SDK v0.22.3.0. ([f045e150])

[8c7724e6]: https://github.com/mpaperno/WASimCommander/commit/8c7724e60ed94e622d5ee2669cf7e000031c2c18
[e16049ac]: https://github.com/mpaperno/WASimCommander/commit/e16049ac69ff15cdcdd9084c7fdab6920a1ffba1
[61a52674]: https://github.com/mpaperno/WASimCommander/commit/61a52674e0dff7e1f3e63ed73a0bed711bb2c479
[17791eef]: https://github.com/mpaperno/WASimCommander/commit/17791eefecc86454c031636a5da9c19d56e21139
[983e7ab6]: https://github.com/mpaperno/WASimCommander/commit/983e7ab609e81af81525ff84431b1c4557447d87
[70e0ef31]: https://github.com/mpaperno/WASimCommander/commit/70e0ef31b01a1a772d9e49102e0a77ec6f3e928b
[90242ed4]: https://github.com/mpaperno/WASimCommander/commit/90242ed494069aba5bcdad839914b9fcfc6521e2
[fe99bbb2]: https://github.com/mpaperno/WASimCommander/commit/fe99bbb25c5dd907e8a4d513769759c4b430580f
[f045e150]: https://github.com/mpaperno/WASimCommander/commit/f045e15007abd6b7b05b97c004a7a55488a33a9b

### WASimClient and WASimClient_CLI (managed .NET)
* Fixed incoming data size check for variable requests which are less than 4 bytes in size. ([c8e74dfa])
* Fixed early timeout being reported on long-running `list()` requests (eg.thousands of L vars). ([a05a28c3])
* Restored ability to specify Unit type for L vars and support for GetCreate with default value/unit and added extra features: ([3090d534], [0a30646d])
  * Added unit name parameter to `setLocalVariable()` and `setOrCreateLocalVariable()`.
  * Added `getOrCreateLocalVariable()`.
  * Added `VariableRequest::createLVar` property.
  * Add optional `create` flag and unit name to `VariableRequest()` c'tor overloads.
* Added async option to `saveDataRequest()` which doesn't wait for server response (`saveDataRequestAsync()` for the C# version). ([82ea4252], [0a30646d])
* Added ability to return a string value with `getVariable()` to make use of new WASimModule feature. ([8e75eb8c], [0e54794b])
* The request updates paused state (set with `setDataRequestsPaused()`) is now saved locally even if not connected to server and will be sent to server upon connection and before sending any queued data requests. 
  This allows connecting and sending queued requests but suspending any actual value checks until needed. ([bea8bccb])
* The `setVariable()` method now verifies that the specified variable type is settable before sending the command to the server. ([576914a2])
* Removed logged version mismatch warning on Ping response.
* Documentation updates.

[c8e74dfa]: https://github.com/mpaperno/WASimCommander/commit/c8e74dfa706647cf785c7e6c811731d8945e49c6
[a05a28c3]: https://github.com/mpaperno/WASimCommander/commit/a05a28c3d1af56444be3fbe54f619e62548736a0
[3090d534]: https://github.com/mpaperno/WASimCommander/commit/3090d5344c3a34c62e81f61237fe1fd91f6b11c5
[0a30646d]: https://github.com/mpaperno/WASimCommander/commit/0a30646d0ae985580d67ed40c8a441a0f5a0ba17
[82ea4252]: https://github.com/mpaperno/WASimCommander/commit/82ea4252bd25423bbeab354799d6be41f053880e
[8e75eb8c]: https://github.com/mpaperno/WASimCommander/commit/8e75eb8c087f5a39fee93c2b7d073500e4f14664
[0e54794b]: https://github.com/mpaperno/WASimCommander/commit/0e54794b2ec8411f42d34a7696426724ffc5e932
[bea8bccb]: https://github.com/mpaperno/WASimCommander/commit/bea8bccba38fae987690d5af259f6f8b22fbc781
[576914a2]: https://github.com/mpaperno/WASimCommander/commit/576914a235c81b73ba0ea85655d913b61cbc5015

### WASimClient_CLI (managed .NET)
* Fixed possible exception when assembling list lookup results dictionary in the off-case of duplicate keys. ([cf46967b])

[cf46967b]: https://github.com/mpaperno/WASimCommander/commit/cf46967b499a9bb19a77a14a47bd2ac29b4d0989

### WASimUI
* Added database of Simulator Variables, Key Events, and Unit types imported from SimConnect SDK online documentation. This is used for:
  * Typing suggestions in the related form fields when entering names of 'A' vars, Key Events, or Unit types.
  * Available as a popup search window from each related form (Variables, Key Events, Data Requests) via button/menu/CTRL-F shortcut.
  * Can be opened as a standalone window for browsing and searching all imported data by type.
* Added ability to import and export Data Requests in _MSFS/SimConnect Touch Portal Plugin_ format with a new editor window available to adjust plugin-specific data before export (category, format, etc.)
* Fixed that the state of current item selections in tables wasn't always properly detected and buttons didn't get enabled/disabled when needed (eg. "Remove Requests" button).
* Added ability to toggle visibility of each main form area of the UI from the View menu (eg. Variables or Key Events groups). Choices are preserved between sessions.
* Simplified the connection/disconnection procedure by providing one action/button for both Sim and Server connections (independent actions still available via extension menu).
* Typing suggestions in combo boxes now use a drop-down menu style selection list by default, and the behavior can be configured independently for each one.
* String type variables can now be used in the "Variables" section for `Get` commands.
* Unit type specifier is now shown and used for 'L' variables as well (unit is optional).
* Added "Get or Create" action/button for 'L' vars.
* The list of 'L' variables loaded from simulator is now sorted alphabetically.
* The Size field in Data Request form is automatically populated with a likely match when a new Unit type is selected.
* Many improvements in table views (all options are saved to user settings and persist between sessions):
  * All column widths are now re-sizable in all tables.
  * Columns can be toggled on/off in the views (r-click for context menu).
  * Can now be sorted by multiple columns (CTRL-click).
  * Option to show filtering (searching) text fields for each column. Filters support wildcards and optional regular expressions.
  * Font size can be adjusted (using context menu or CTRL key with `+`, `-`, or `0` to reset.
  * Tooltips shown with data values when hovered over table cells (readable even if text is too long to fit in the column).
* Numerous shortcuts and context menus added throughout, each relevant to the respective forms/tables currently being used or clicked.
* Last selected variable types and data request type are saved between sessions.
* Most actions/buttons which require a server connection to work are now disabled when not connected.
* When loading data requests from a file while connected to the server, the requests are now sent asynchronously, improving UI responsiveness.
* More minor quality-of-life improvements!

**Full log:** [v1.1.2.0...1.2.0.0](https://github.com/mpaperno/WASimCommander/compare/1.1.2.0...1.2.0.0)

----
The release contains several download options (the actual file names have version numbers appended):
- `WASimCommander_SDK` - All header files, pre-built static, dynamic, and managed (.NET) libs, packaged WASM module, pre-build GUI, reference documentation, and other tools/examples.
- `WASimModule` - Just the WASM module component, ready to be dropped into a MSFS _Community_ folder (so that the folder _wasimcommander-module_ is inside the _Community_ folder). MSFS would need to be restarted after adding the module for the first time.
- `WASimUI` - Just the GUI application, which is very handy in combination with the WASM module for exploring the system. No installer, simply unzip and run the _WASimUI.exe_ file from either Explorer or a command line.

<hr /><em>This discussion was created from the release <a href='https://github.com/mpaperno/WASimCommander/releases/tag/1.2.0.0'>v1.2.0 - Large update with fixes and new features</a>.</em>

**Comment by dinther** [comment]
Hi Max, been away from my project for a while but I am back. I read through your updates and updated my dev code. All is well.

Nice to see you can do now getVariable for strings so I can remove my calc workaround in my wrapper.

That of course leads to the question... Setting strings with setVariable. That is just not possible with WASimCommander at all right now correct?

**Comment by mpaperno** [comment]
Hi Paul, good to hear from you.

That's right, currently `setVariable()` can't handle strings.  Though I'm considering adding that, since it's "unbalanced" (or unexpected) right now the way it is.  Basically it would just need to convert to calculator code internally (probably on the client side), probably what you're doing now with the workaround (and what I currently do as well).

I've been avoiding adding client-specific functionality to the client (stuff that's not directly related to the "WASim API"), but I think Hans has convinced me to add custom named Key Event handling in the client anyway... so that kinda opens the door to other abstractions as well. :)

Cheers,
-Max

**Reply by dinther** [reply]
Oh cool! Hans Billiet? His excellent posts on the forum are the reason I ventured down this path. I convinced him to look at your project and then I went radio silent and left him hanging.

It troubles me to see so little interest for your project. This effort ought to be a huge community driven open source project. It pays off in so may ways. I destroyed and made a new SimCom project (Github drives me nuts)

I am currently considering to make all the variables unitless because currently
`(A:AUTOPILOT HEADING LOCK DIR,degrees)` and `(A:AUTOPILOT HEADING LOCK DIR,radians)` would be two separate SimVal objects in my code.

I think the client should be responsible for the unit conversion.

**Reply by mpaperno** [reply]
Yep, that very same Hans.  We've been having a lively e-mail discussion (torrent?).  I tried to ping him on here but can't figure out his GH user handle.  He did mention the connection with you, and thank you for bringing it to his attention!

It's definitely nice to get feedback and ideas on the project, finally. Guess I'm not a big promoter or "community builder."  :)

It's certainly convenient to get SimConnect to do all the unit conversions. But indeed likely not very efficient to request the same variable multiple times just for that convenience.  That could be a good feature for an abstraction layer, where all variable requests go into a pool, and the manager is responsible for finding duplicates and doing unit conversions (and perhaps formatting) per requestor as needed.  That would possibly require a significant unit conversion exercise if it were to be fully adaptable for any variable.  But it's a nice idea... something I could actually use in the Touch Portal plugin I maintain.

**Reply by dinther** [reply]
Hi Max, I have been fighting a bit with my Visual Studio 2022 a bit. I really wanted to be able to compile SimCom for .NET6 but it kept failing for a lot of reasons. But I finally figured it out. However, one that is important for your documentation (Unless I missed it) is that the user manually needs to copy `C:\MSFS SDK\WASimCommander\lib\managed\Ijwhost.dll` to the output directory just like is advised with simconnect.dll.

**Reply by HansBilliet** [reply]
Hello guys,

Nice to see you both together in my inbox!

@dinther I did figure that jihost.dll-stuff out thanks to one of my posts on Stackexchange. Indeed something you have to know (now I know).

Ah, I didn't realize you couldn't set strings with SetVariable. But indeed, only double for the time being.

I had a lot of conversations with Max (I call it "letters from Max"). I have been diving in his code, and I learned a lot! As you probably know, I'm rewriting my tool (the one I described in my posts) by using his WASimCommander. Reason is that his WASM contains the few things that I was still missing - and looking at the concept, it's more robust then mine. Because I need it to connect with my own made cockpit hardware, I needed also to be able to control Custom Events, which is not yet implemented. But I took the challenge to implement it myself (with some help of Max of course). I might do this this weekend. After that, I can start with the real integration.

I am more of a C# programmer, but I start appreciating C++ a lot. Although, it would be too much of a challenge to completely write my tool in C++, so will still have to deal with the WASimClient_CLI.

@mpaperno Just realizing this is a public GitHub channel, right? Wouldn't it be better to just use email? Although, I don't think @dinther  likes that so much :-)
